text/0000-unnamed-fields.md

```diff
@@ -0,0 +1,350 @@
+        e: f32,
+    },
+    f: u64,
+}
+```
+
+This structure contains six fields: `a`, `b`, `c`, `d`, `e`, and `f`. Safe code
+can access fields `a` and `f`, at any time, since those fields do not lie
+within a union and do not overlap with any other field. Unsafe code can access
+the remaining fields. This definition effectively acts as the overlap of the
+following three structures:
+
+```rust
+// variant 1
+struct S {
+    a: u32,
+    b: u32,
+    f: u64,
+}
+
+// variant 2
+struct S {
+    a: u32,
+    c: u16,
+    d: f16,
+    f: u64,
+}
+
+// variant 3
+struct S {
+    a: u32,
+    e: f32,
+    f: u64,
+}
+```
+
+## Instantiation
+
+Given the following declaration:
+
+```rust
+struct S {
+    a: u32,
+    union {
+        b: u32,
+        struct {
+            c: u16,
+            d: f16,
+        },
+        e: f32,
+    },
+    f: u64,
+}
+```
+
+All of the following will instantiate a value of type `S`:
+
+- `S { a: 1, b: 2, f: 3.0 }`
+- `S { a: 1, c: 2, d: 3.0, f: 4.0 }`
+- `S { a: 1, e: 2.0, f: 3.0 }`
+
+## Representation
+
+By default, Rust lays out structures using its native representation,
+`repr(Rust)`; that representation permits any layout that can store all the
+non-overlapping fields simultaneously, and makes no other guarantees about the
+storage of unnamed fields.
+
+When using this mechanism to define a C interface, remember to use the
+`repr(C)` attribute to match C's data structure layout. Any representation
+attribute applied to the top-level structure also applies to every unnamed
+field within that declaration. Such a structure defined with `repr(C)` will use
+a representation identical to the same structure with all unnamed fields
+transformed to equivalent named fields of a struct or union type with the same
+fields.
+
+## Derive
+
+A `struct` or `union` containing unnamed fields may derive `Copy`, `Clone`, or
+both, if all the fields it contains (including within unnamed fields) also
+implement `Copy`.
+
+A `struct` containing unnamed fields may derive `Clone` if every field
+contained directly in the `struct` implements `Clone`, and every field
+contained within an unnamed `union` (directly or indirectly) implements `Copy`.
+
+## Ambiguous field names
+
+You cannot use this feature to define multiple fields with the same name. For
+instance, the following definition will produce an error:
+
+```rust
+struct S {
+    a: u32,
+    union {
+        a: u32,
+        b: f32,
+    },
+}
+```
+
+The error will identify the duplicate `a` fields as the sources of the error.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Parsing
+
+Within a struct's fields, in place of a field name and value, allow
+`union { fields }`, where `fields` allows everything allowed within a `union`
+declaration. Conversely, within a union's fields, in place of a field name
+and value, allow `struct { fields }`, where `fields` allows everything allowed
+within a `struct` declaration.
+
```

Comment:
The fields are not inlined for the sake of layout, they are only "inlined" in terms of the syntax you use to refer to the fields. Because this RFC explicitly only works with `#[repr(C)]` it must necessarily match what C does, so there's no decisions to be made about which layout to use.